File: src/TelecommsSimulation.Core.CS/Phone.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Threading.Tasks;

using TelecommsSimulation.Utilities;

namespace TelecommsSimulation.Core;

public class Phone : ILocatableMovable, INotifyPropertyChanged
{
    public string Number { get; }

    private Coordinates _location;

    public Coordinates Location
    {
        get => _location;
        set
        {
            SetField(ref _location, value);
            SwitchBaseStationIfNeeded();
        }
    }

    private PhoneState _state = PhoneState.Off;

    public PhoneState State
    {
        get => _state;
        private set => SetField(ref _state, value);
    }

    private CallState _callState = CallState.Idle;

    public CallState CallState
    {
        get => _callState;
        private set => SetField(ref _callState, value);
    }

    private string? _inCallWith;

    public string? InCallWith
    {
        get => _inCallWith;
        private set => SetField(ref _inCallWith, value);
    }

    private double _batteryLevel;

    public double BatteryLevel
    {
        get => _batteryLevel;
        set => SetField(ref _batteryLevel, MathExtensions.Clamp(value, 0, 100));
    }

    protected BaseStation? ConnectedBaseStation;

    public string? ConnectedBaseStationName => ConnectedBaseStation?.Name;

    protected readonly SynchronizationContext? UiSynchronizationContext;

    protected TaskCompletionSource<CallAnswer>? CallAnswerCompletionSource;

    public event EventHandler? Connected;
    public event EventHandler? Disconnected;
    public event AsyncEventHandler<CallEventArgs>? CallReceived;
    public event EventHandler<CallEventArgs>? CallEnded;

    public Phone(string number, Coordinates location, double batteryLevel = 100)
    {
        UiSynchronizationContext = SynchronizationContext.Current;

        Number = number;
        Location = location;
        BatteryLevel = batteryLevel;
    }

    public virtual void TurnOn()
    {
        if (State is not PhoneState.Off)
            throw new InvalidPhoneStateException();

        if (BatteryLevel == 0)
            return;

        State = PhoneState.Disconnected;
    }

    public virtual void TurnOff()
    {
        if (State is PhoneState.Off)
            throw new InvalidPhoneStateException();

        if (State is PhoneState.Connected)
            DisconnectFromBaseStation();

        State = PhoneState.Off;
    }

    public virtual void ConnectToBaseStation()
    {
        switch (State)
        {
            case PhoneState.Off:
            {
                throw new InvalidPhoneStateException();
            }
            case PhoneState.Connected:
            {
                DisconnectFromBaseStation();
                break;
            }
        }

        BaseStation? nearestStation = FindBaseStation();

        if (nearestStation is null)
            return;

        nearestStation.RegisterPhone(this);
        ConnectedBaseStation = nearestStation;
        OnPropertyChanged(nameof(ConnectedBaseStationName));

        State = PhoneState.Connected;
        Connected?.Invoke(this, EventArgs.Empty);
    }

    public virtual void DisconnectFromBaseStation()
    {
        if (State is not PhoneState.Connected)
            throw new InvalidPhoneStateException();

        if (CallState is not CallState.Idle)
            EndCall();

        ConnectedBaseStation!.UnregisterPhone(this);
        ConnectedBaseStation = null;
        OnPropertyChanged(nameof(ConnectedBaseStationName));

        State = PhoneState.Disconnected;
        Disconnected?.Invoke(this, EventArgs.Empty);
    }

    public virtual void SwitchBaseStationIfNeeded()
    {
        if (State is not PhoneState.Connected)
            return;

        if (ConnectedBaseStation!.IsEnabled && IsCoveredBy(ConnectedBaseStation!))
            return;

        BaseStation? nearestStation = FindBaseStation();

        if (nearestStation is null)
        {
            DisconnectFromBaseStation();
            return;
        }

        ConnectedBaseStation!.UnregisterPhone(this);

        nearestStation.RegisterPhone(this);

        ConnectedBaseStation = nearestStation;
        OnPropertyChanged(nameof(ConnectedBaseStationName));
    }

    public virtual async Task<CallResult> MakeCall(string to)
    {
        if (State is not PhoneState.Connected)
            throw new InvalidPhoneStateException();

        if (CallState is not CallState.Idle)
            throw new InvalidCallStateException();

        InCallWith = to;
        CallState = CallState.Dialing;

        CallResult result = await ConnectedBaseStation!.InitiateCall(this, to);

        if (result is not CallResult.Established)
        {
            InCallWith = null;
            CallState = CallState.Idle;

            return result;
        }

        CallState = CallState.InCall;

        return result;
    }

    public virtual Task<CallAnswer> ReceiveCall(string from)
    {
        if (State is not PhoneState.Connected)
            throw new InvalidPhoneStateException();

        if (CallState is not CallState.Idle)
            throw new InvalidCallStateException();

        InCallWith = from;
        CallState = CallState.AwaitingAnswer;

        TaskCompletionSource<CallAnswer> callAnswerCompletionSource = CallAnswerCompletionSource =
            new TaskCompletionSource<CallAnswer>(TaskCreationOptions.RunContinuationsAsynchronously);

        CallReceived?.Invoke(this, new CallEventArgs(from, Number));

        return callAnswerCompletionSource.Task;
    }

    public virtual void AnswerCall(CallAnswer answer)
    {
        if (State is not PhoneState.Connected)
            throw new InvalidPhoneStateException();

        if (CallState is not CallState.AwaitingAnswer)
            throw new InvalidCallStateException();

        CallAnswerCompletionSource!.SetResult(answer);

        CallAnswerCompletionSource = null;

        if (answer is CallAnswer.Agreed)
        {
            CallState = CallState.InCall;
            return;
        }

        CallState = CallState.Idle;
        InCallWith = null;
    }

    public virtual void EndCall()
    {
        if (State is not PhoneState.Connected)
            throw new InvalidPhoneStateException();

        if (CallState is CallState.Idle)
            throw new InvalidCallStateException();

        CallAnswerCompletionSource?.SetCanceled();
        CallAnswerCompletionSource = null;

        CallEnded?.Invoke(this, new CallEventArgs(Number, InCallWith!));

        InCallWith = null;
        CallState = CallState.Idle;
    }

    protected virtual BaseStation? FindBaseStation()
    {
        BaseStation? nearestStation =
            BaseStation.AllEnabledStations
                .Where(IsCoveredBy)
                .MinBy(x => x.Location.CalculateKmDistanceTo(Location));

        return nearestStation;
    }

    protected bool IsCoveredBy(BaseStation baseStation) =>
        baseStation.Location.CalculateKmDistanceTo(Location) <= baseStation.CoverageRadiusKm;

    public event PropertyChangedEventHandler? PropertyChanged;

    protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        // Workaround: WinForms' lack of threading awareness and the challenge of re-invoking handlers in C++/CLI (i.e. no support for lambda-functions)
        if (UiSynchronizationContext != null)
            UiSynchronizationContext.Post(_ => PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName)), null);
        else
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }

    protected bool SetField<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value))
            return false;

        field = value;

        OnPropertyChanged(propertyName);

        return true;
    }
}

File: src/TelecommsSimulation.Core/Phone.cpp
#include "pch.h"

#include "Phone.h"

#include "BaseStation.h"
#include "InvalidCallStateException.h"
#include "InvalidPhoneStateException.h"
#include "LinqExtensions.h"

using namespace System;
using namespace System::Collections::Generic;
using namespace System::ComponentModel;
using namespace System::Linq;
using namespace System::Threading;
using namespace System::Threading::Tasks;

using namespace TelecommsSimulation::Utilities;

namespace TelecommsSimulation::Core
{
    String^ Phone::Number::get()
    {
        return _number;
    }

    void Phone::Number::set(String^ value)
    {
        SetField<String^>(_number, value, "Number");
    }

    Coordinates Phone::Location::get()
    {
        return _location;
    }

    void Phone::Location::set(Coordinates value)
    {
        SetField<Coordinates>(_location, value, "Location");
        SwitchBaseStationIfNeeded();
    }

    PhoneState Phone::State::get()
    {
        return _state;
    }

    void Phone::State::set(PhoneState value)
    {
        SetField<PhoneState>(_state, value, "State");
    }

    CallState Phone::CallState::get()
    {
        return _callState;
    }

    void Phone::CallState::set(Core::CallState value)
    {
        SetField<Core::CallState>(_callState, value, "CallState");
    }

    String^ Phone::InCallWith::get()
    {
        return _inCallWith;
    }

    void Phone::InCallWith::set(String^ value)
    {
        SetField<String^>(_inCallWith, value, "InCallWith");
    }

    double Phone::BatteryLevel::get()
    {
        return _batteryLevel;
    }

    void Phone::BatteryLevel::set(double value)
    {
        SetField<double>(_batteryLevel, MathExtensions::Clamp(value, 0, 100), "BatteryLevel");
    }

    String^ Phone::ConnectedBaseStationName::get()
    {
        return ConnectedBaseStation->Name;
    }

    Phone::Phone(String^ number, Coordinates location) : Phone(number, location, 100) {}

    Phone::Phone(String^ number, Coordinates location, double batteryLevel)
    {
        UiSynchronizationContext = SynchronizationContext::Current;

        Number = number;
        Location = location;
        BatteryLevel = batteryLevel;
    }

    void Phone::TurnOn()
    {
        if (State != PhoneState::Off)
            throw gcnew InvalidPhoneStateException();

        if (BatteryLevel == 0)
            return;

        State = PhoneState::Disconnected;
    }

    void Phone::TurnOff()
    {
        if (State == PhoneState::Off)
            throw gcnew InvalidPhoneStateException();

        if (State == PhoneState::Connected)
            DisconnectFromBaseStation();

        State = PhoneState::Off;
    }

    void Phone::ConnectToBaseStation()
    {
        switch (State)
        {
            case PhoneState::Off:
                {
                    throw gcnew InvalidPhoneStateException();
                }
            case PhoneState::Connected:
                {
                    DisconnectFromBaseStation();
                    break;
                }
            default:
                break;
        }

        BaseStation^ nearestStation = FindBaseStation();

        if (nearestStation == nullptr)
            return;

        nearestStation->RegisterPhone(this);
        ConnectedBaseStation = nearestStation;
        OnPropertyChanged("ConnectedBaseStationName");

        State = PhoneState::Connected;
        Connected(this, EventArgs::Empty);
    }

    void Phone::DisconnectFromBaseStation()
    {
        if (State != PhoneState::Connected)
            throw gcnew InvalidPhoneStateException();

        if (CallState != Core::CallState::Idle)
            EndCall();

        ConnectedBaseStation->UnregisterPhone(this);
        ConnectedBaseStation = nullptr;
        OnPropertyChanged("ConnectedBaseStationName");

        State = PhoneState::Disconnected;
        Disconnected(this, EventArgs::Empty);
    }

    void Phone::SwitchBaseStationIfNeeded()
    {
        if (State != PhoneState::Connected)
            return;

        if (ConnectedBaseStation->IsEnabled && IsCoveredBy(ConnectedBaseStation))
            return;

        BaseStation^ nearestStation = FindBaseStation();

        if (nearestStation == nullptr)
        {
            DisconnectFromBaseStation();
            return;
        }

        ConnectedBaseStation->UnregisterPhone(this);

        nearestStation->RegisterPhone(this);

        ConnectedBaseStation = nearestStation;
        OnPropertyChanged("ConnectedBaseStationName");
    }

    Task<CallResult>^ Phone::MakeCall(String^ to)
    {
        if (State != PhoneState::Connected)
            throw gcnew InvalidPhoneStateException();

        if (CallState != Core::CallState::Idle)
            throw gcnew InvalidCallStateException();

        InCallWith = to;
        CallState = Core::CallState::Dialing;

        return ConnectedBaseStation->InitiateCall(this, to)->ContinueWith(
            gcnew Func<Task<CallResult>^, CallResult>(this, &Phone::MakeCallContinuation),
            TaskContinuationOptions::RunContinuationsAsynchronously);
    }

    CallResult Phone::MakeCallContinuation(Task<CallResult>^ task)
    {
        const CallResult result = task->Result;

        if (result != CallResult::Established)
        {
            InCallWith = nullptr;
            CallState = Core::CallState::Idle;

            return result;
        }

        CallState = Core::CallState::InCall;

        return result;
    }

    Task<CallAnswer>^ Phone::ReceiveCall(String^ from)
    {
        if (State != PhoneState::Connected)
            throw gcnew InvalidPhoneStateException();

        if (CallState != Core::CallState::Idle)
            throw gcnew InvalidCallStateException();

        InCallWith = from;
        CallState = Core::CallState::AwaitingAnswer;

        TaskCompletionSource<CallAnswer>^ callAnswerCompletionSource = CallAnswerCompletionSource =
            gcnew TaskCompletionSource<CallAnswer>(TaskCreationOptions::RunContinuationsAsynchronously);

        CallReceived(this, gcnew CallEventArgs(from, Number));

        return callAnswerCompletionSource->Task;
    }

    void Phone::AnswerCall(CallAnswer answer)
    {
        if (State != PhoneState::Connected)
            throw gcnew InvalidPhoneStateException();

        if (CallState != Core::CallState::AwaitingAnswer)
            throw gcnew InvalidCallStateException();

        CallAnswerCompletionSource->SetResult(answer);
        CallAnswerCompletionSource = nullptr;

        if (answer == CallAnswer::Agreed)
        {
            CallState = Core::CallState::InCall;
            return;
        }

        CallState = Core::CallState::Idle;
        InCallWith = nullptr;
    }

    void Phone::EndCall()
    {
        if (State != PhoneState::Connected)
            throw gcnew InvalidPhoneStateException();

        if (CallState == Core::CallState::Idle)
            throw gcnew InvalidCallStateException();

        if (CallAnswerCompletionSource != nullptr)
            CallAnswerCompletionSource->SetCanceled();
        CallAnswerCompletionSource = nullptr;

        CallEnded(this, gcnew CallEventArgs(Number, InCallWith));

        InCallWith = nullptr;
        CallState = Core::CallState::Idle;
    }

    BaseStation^ Phone::FindBaseStation()
    {
        return LinqExtensions::MinBy(
            Enumerable::Where(
                BaseStation::AllEnabledStations,
                gcnew Func<BaseStation^, bool>(this, &Phone::IsCoveredBy)),
            gcnew Func<BaseStation^, double>(this, &Phone::FindBaseStationMinByLambda));
    }

    double Phone::FindBaseStationMinByLambda(BaseStation^ x)
    {
        return x->Location.CalculateKmDistanceTo(Location);
    }

    bool Phone::IsCoveredBy(BaseStation^ baseStation)
    {
        return baseStation->Location.CalculateKmDistanceTo(Location) <= baseStation->CoverageRadiusKm;
    }

    void Phone::OnPropertyChanged(String^ propertyName)
    {
        // Workaround: WinForms' lack of threading awareness and the challenge of re-invoking handlers in C++/CLI
        if (UiSynchronizationContext != nullptr)
        {
            UiSynchronizationContext->Post(
                gcnew SendOrPostCallback(this, &Phone::OnPropertyChangedInvoke),
                propertyName);
        }
        else
        {
            PropertyChanged(this, gcnew PropertyChangedEventArgs(propertyName));
        }
    }

    void Phone::OnPropertyChangedInvoke(Object^ propertyName)
    {
        Phone::OnPropertyChanged(safe_cast<String^>(propertyName));
    }

    generic<typename T>
    bool Phone::SetField(T% field, T value, String^ propertyName)
    {
        if (EqualityComparer<T>::Default->Equals(field, value))
            return false;

        field = value;
        OnPropertyChanged(propertyName);
        return true;
    }
}
